src/connectors/impls/gpubsub/producer.rs

```diff
@@ -0,0 +1,284 @@
+        #[cfg(test)]
+        {
+            skip_authentication = self.config.skip_authentication;
+        }
+
+        let client = if skip_authentication {
+            info!("Skipping auth...");
+
+            PublisherClient::with_interceptor(
+                channel,
+                AuthInterceptor {
+                    token: Box::new(|| Ok(Arc::new(String::new()))),
+                },
+            )
+        } else {
+            let token = Token::new()?;
+
+            PublisherClient::with_interceptor(
+                channel,
+                AuthInterceptor {
+                    token: Box::new(move || {
+                        token.header_value().map_err(|_| {
+                            Status::unavailable("Failed to retrieve authentication token.")
+                        })
+                    }),
+                },
+            )
+        };
+
+        self.client = Some(client);
+
+        Ok(true)
+    }
+
+    async fn on_event(
+        &mut self,
+        _input: &str,
+        event: Event,
+        ctx: &SinkContext,
+        serializer: &mut EventSerializer,
+        _start: u64,
+    ) -> Result<SinkReply> {
+        let client = self.client.as_mut().ok_or(ErrorKind::ClientNotAvailable(
+            "PubSub",
+            "The publisher is not connected",
+        ))?;
+
+        let mut messages = Vec::with_capacity(event.len());
+
+        for (value, meta) in event.value_meta_iter() {
+            for payload in serializer.serialize(value, event.ingest_ns)? {
+                let ordering_key = meta
+                    .get("gpubsub_producer")
+                    .get("ordering_key")
+                    .as_str()
+                    .map_or_else(|| "".to_string(), ToString::to_string);
+
+                messages.push(PubsubMessage {
+                    data: payload,
+                    attributes: HashMap::new(),
+                    // publish_time and message_id will be ignored in the request and set by server
+                    message_id: "".to_string(),
+                    publish_time: None,
+
+                    // fixme get from the metadata
+                    ordering_key,
+                });
+            }
+        }
+
+        if let Ok(inner_result) = client
+            .publish(PublishRequest {
+                topic: self.config.topic.clone(),
+                messages,
+            })
+            .timeout(Duration::from_nanos(self.config.request_timeout))
+            .await
+        {
+            if let Err(error) = inner_result {
+                error!("Failed to publish a message: {}", error);
+                ctx.swallow_err(
+                    ctx.notifier.connection_lost().await,
+                    "Failed to notify about PubSub connection loss",
+                );
+
+                Ok(SinkReply::FAIL)
+            } else {
+                Ok(SinkReply::ACK)
+            }
+        } else {
+            ctx.swallow_err(
```

Comment:
Do we always want to assume a lost connection if we hit a timeout? Or should we simply fail? Calling the `connection_lost()` method will have the effect that the circuit breaker will be triggered and we will attempt a reconnect. If we just fail on the other hand, and we hit repeated timeouts, we might never reconnect.

Maybe we should discuss a golden strategy here that we can/should apply to all sink-connectors?

We do have the `qos::backpressure` and `qos::percentile` operators who can be used to either stop upstreams from sending using the circuit-breaker or just drop some messages.

Reply:
I would say we should just `FAIL` here and not call `connection_lost`. But pls leave a comment that this might need fixing.

Reply:
this is the last thing that needs fixing